boards/microduino-corerf: Initial support

benpicco opened the pull request:
### Contribution description
The [Microduino CoreRF](https://wiki.microduinoinc.com/Microduino-Module_CoreRF) is an inexpensive breakout board for the [ATmega128RFA1](https://www.microchip.com/wwwproducts/en/ATmega128RFA1) MCU.

The ATmega128RFA1 is the predecessor of the ATmega256RFR2. The ATmega256/128/64RFR2 data sheet says

> Backward compatibility of the ATmega256/128/64RFR2 to the ATmega128RFA1 is
provided in most cases. However some incompatibilities may exist.
> The ATmega256/128/64RFR2 uses the same package as the ATmega128RFA1.

Further search in the data sheet shows that the incompatibilities only concern different ADC characteristics.

### Testing procedure

The board itself is rather spartan. It only provides a reset button and a non-controllable power LED.

The RX/TX pins have to be connected to a user-provided serial adapter to provide a UART interface.

#### Flashing

The board comes pre-flashed with an Arduino compatible bootloader, but I managed to delete it.
I've then spent way too much time trying to re-flash it (especially since most instructions call for a special capacitor on the reset line) until I discovered that flashing works remarkably well with a FT232H USB dongle I happened to have - no extra components needed.

All guides on how to re-flash the Arduino bootloader recommend using the Arduino IDE - which does not seem to support the FT232H dongle.

Since it works well with plain `avrdude` I decided not to bother with the bootloader and instead enjoy the additional ROM space.

If required, I can see how I can restore the bootloader and flash using that.

##### Tests
Since the board is basically just a plain Atmega MCU, I just checked if stdio and timers work.

For this I flashed `examples/default`, `tests/xtimer_msg` and `tests/xtimer_drift` - they all behaved as expected.

I did not test the radio yet.

### Issues/PRs references
#9172 will add support for the radio

maribu commented on boards/common/atmega/board.c:
```diff
@@ -29,6 +29,8 @@
 
 void led_init(void);
 
+void __attribute__((weak)) led_init(void) {}
```
This is nice. But I think not everyone is aware what `__attribute__((weak))` means. Mind to add a short comment, maybe something like "Add an empty led_init as fall back: if at link time another implementation of led_init not marked as weak (a.k.a. a strong symbol) is present, it will be linked in instead."

benpicco replied:
Ok, added the comment.

maribu commented on boards/microduino-corerf/doc.txt:
```diff
@@ -0,0 +1,80 @@
+| I2Cs		 | 1 (called TWI)	 |
+| Vcc		 | 1.8V - 3.6V		 |
+| Datasheet / Reference Manual | [Datasheet and Reference Manual](http://ww1.microchip.com/downloads/en/DeviceDoc/Atmel-8266-MCU_Wireless-ATmega128RFA1_Datasheet.pdf) |
+| Board Manual	 | [Wiki Page](https://wiki.microduinoinc.com/Microduino-Module_CoreRF) |
```
I now this is super bike-shedding and jerky to ask for, but I somehow find unaligned table columns super irritating. Would you mind to align the columns? (No need to align the right column of the last two rows with the long links.)

maribu commented on cpu/atmega128rfa1/include/atmega_pcint.h:
```diff
@@ -0,0 +1,15 @@
+#endif
+
+#define ATMEGA_PCINT_MAP_PCINT0 GPIO_PIN(PORT_B, 0), GPIO_PIN(PORT_B, 1), GPIO_PIN(PORT_B, 2), GPIO_PIN(PORT_B, 3), GPIO_PIN(PORT_B, 4), GPIO_PIN(PORT_B, 5), GPIO_PIN(PORT_B, 6), GPIO_PIN(PORT_B, 7)
+#define ATMEGA_PCINT_MAP_PCINT1 GPIO_PIN(PORT_E, 0), GPIO_UNDEF, GPIO_UNDEF, GPIO_UNDEF, GPIO_UNDEF, GPIO_UNDEF, GPIO_UNDEF, GPIO_UNDEF
```
I didn't thought this through, but maybe PCINT1 can be treated like an EXTINT if there is only one pin triggering the interrupt? If this would be possible to do with little effort and code spend on it, this should be more efficient.

benpicco replied:
TBH, I just copied that from `cpu/atmega256rfr2/include/atmega_pcint.h`.

Looking at the data sheet confirms that `PE0` is the only `PCINT` on this port.
Is there already a define for what you suggest, or would that require changes to `atmega_common/periph/gpio.c`? If so, I'd rather do that in a later PR.

benpicco replied:
Looking at it, it seems like `CPU_ATMEGA_EXT_INTS` is then wrong for both - it should be the same pins as the ones defined in `ATMEGA_PCINT_MAP_PCINTx`, right?

maribu replied:
No, those are two different types of interrupts. The EXT_INTS have one interrupt vector per pin, the PCINT have one interrupt vector per group of up to 8 pins. So for PCINT the ISR has to figure out which pin triggered the interrupt, if that info is needed. (For compliance with the RIOT API is is always needed.)

maribu commented on cpu/atmega128rfa1/ldscripts_compat/avr_2.26.ld:
```diff
@@ -0,0 +1,262 @@
+__FUSE_REGION_LENGTH__ = DEFINED(__FUSE_REGION_LENGTH__) ? __FUSE_REGION_LENGTH__ : 1K;
+__LOCK_REGION_LENGTH__ = DEFINED(__LOCK_REGION_LENGTH__) ? __LOCK_REGION_LENGTH__ : 1K;
+__SIGNATURE_REGION_LENGTH__ = DEFINED(__SIGNATURE_REGION_LENGTH__) ? __SIGNATURE_REGION_LENGTH__ : 1K;
+__USER_SIGNATURE_REGION_LENGTH__ = DEFINED(__USER_SIGNATURE_REGION_LENGTH__) ? __USER_SIGNATURE_REGION_LENGTH__ : 1K;
```
This part is the only part specific to the CPU, right? So the rest of this file is identical with the other AVR linker scripts, and I can skip that in the review, right?

benpicco replied:
Yes, and like the other ones it was just generated with `avr-gcc -Wl,--verbose -mmcu=atmega128rfa1`.

It's only there for compatibility with avr-binutils older than 2.26

maribu replied:
> it was just generated with `avr-gcc -Wl,--verbose -mmcu=atmega128rfa1`.

Even better, I don't need to check at all :-)

maribu commented on boards/microduino-corerf/doc.txt:
```diff
@@ -0,0 +1,80 @@
+|  3.3V  | 3.3V              |
+|  GND   | GND               |
+
+Now you can simply type
```
Please add a big warning that the GPIOs are not 5V tolerant, unlike pretty much all other ATmegas. (Luckily I checked this before using my 5V ISP, so I will still be able to test this PR :-))

maribu commented:
OK, the Arduino-Mega2560 still works fine. The test in `tests/periph_cpuid` works as expected on the microduino

maribu commented on boards/microduino-corerf/Makefile.include:
```diff
@@ -0,0 +1,16 @@
+
+# PROGRAMMER defaults to UM232H which is a FT232H breakout board
+# externally connected using wires
+PROGRAMMER ?= UM232H
```
If you want to me super happy, you can declare this like that:

``` Makefile
PROGRAMMER_MICRODUINO_CORERF ?= UM232H
PROGRAMMER ?= $(PROGRAMMER_MICRODUINO_CORERF)
```

But that is an optional suggestion. (Squash right away, if you decide to do this.)

benpicco replied:
Of course I want you to be happy - but what is the purpose of this?

maribu replied:
I could add `export PROGRAMMER_MICRODUINO_CORERF=dragon_isp` to my `.profile`. If I would add `PROGRAMMER=dragon_isp`, I would have trouble flashing ARM boards with a configurable programmer.

benpicco commented:
I'll still need to update some `BOARD_INSUFFICIENT_MEMORY` lists. But some test builds will fail already because of #12412 :wink:

benpicco commented:
I added the board to the `BOARD_INSUFFICIENT_MEMORY` lists. That revealed some more issues in Murdock for which I had to provide more !fixups.

Are you fine with those? Than I'll squash again.

maribu commented:
Yes. please squash again. I'd like to test one final time before merging.

(@maribu: Note to myself: Add `boards/common/atmega/Makefile.features` to reduce code duplication.)

benpicco commented:
Rebased now that #12484 is merged.
#12485 eased the pain: I added the `Makefile.ci` target and then ran `for i in *; do make -C $i Makefile.ci; done`

This would generate fresh `Makefile.ci`s based on the content of the old `Makefile`. `find . -name Makefile.ci -size 37c -delete` then helped to get rid of the empty ones.

Merging the `Makefile.ci`s was then much easier than going through everything by hand again.

maribu commented:
The very second I wrote this, Murdock passed :rofl:

benpicco commented:
Thank you for the review!
And it sounds like you even bought one of those devices yourself 😃

maribu commented:
Two. I figured I sooner or later someone should test the driver for the 802.15.4 transceiver, and that will be more fun with two devices. (Even though I also have other boards with radios that should be compatible over the air.)

aabadie commented:
It seems 2 unrelated commits went through: d92e2b25ac1ac5c6961b43d6fccd2a2dc661ae79 and 252f22bfe29b8a556deaad8210e891d763620d81. They are harmless, so not a big deal, but we should be careful that it doesn't happen too often.

maribu commented:
They were needed to resolve name conflicts

maribu commented:
The commit messages also document this. So the commit were required to resolve compilation issues, relevant for this PR (as they resolve compilation issues), reasonable (only internal defines are changes and the bew names make sense) and properly documented (the commit messages are right to the point).

aabadie commented:
Indeed, I missed the detailed part of the commit messages. At first sight, this was surprising.
